chore: update data node operator documentation (#556)
vegaprotocol/documentation: docs/node-operators/get-started/setup-datanode.md

    -e POSTGRES_PASSWORD=DATABASE_PASSWORD \
    -e POSTGRES_DB=DATABASE_NAME \
    -p LOCALDB_PORT:5432

>     -v /host_path/to/snapshotsCopyTo:/host_path/to/snapshotsCopyTo:z \

Author: Resolved https://github.com/vegaprotocol/documentation/pull/556/commits/7c42130487b2326cabd0e9470ba7f297376f173a

- `database_name` is the name of the database you want to use for storing the data.
- `localdb_port` is the port you want to use to connect to the database on your local machine. (5432 is the default port for Postgresql database server and may not be available if you already have a postgresql database server running on your machine and want to use Docker for testing).
- `/host_path/to/snapshotsCopyTo` is the path on your host machine where you want to store the snapshots that are generated by data node. Snapshots allow you to restore the database to a previous state. By default, this folder will be in the `state/data-node/networkhistory/snapshotsCopyTo` folder under the `--home` folder you set for data node. [See Data Node Configuration](#data-node-configuration)
- `/host_path/to/snapshotsCopyFrom` is the path on your host machine where you want to store the snapshots that are retrieved from IPFS and can be used to rebuild a data node database from another data node. By default, this folder will be in the `state/data-node/networkhistory/snapshotsCopyFrom` folder under the `--home` folder you set for data node. [See Data Node Configuration](#data-node-configuration)

> :::note Snapshot paths
> If you intend to use the Docker version of TimescaleDB and you do not mount the `snapshotsCopyTo` and `snapshotsCopyFrom` paths to the TimescaleDB container, Data Node will panic when trying to start from network history because TimescaleDB is not able to access the network history files Data Node retrieves and tries to import. The mounted path in the container must match the path of the `snapshotsCopyTo` and `snapshotsCopyFrom` paths on the Docker host.
> :::

Reviewer: ```suggestion
:::note Snapshot paths
If you intend to use the Docker version of TimescaleDB, the mounted path in the container must match the path of the `snapshotsCopyTo` and `snapshotsCopyFrom` paths on the Docker host. If you do not mount the `snapshotsCopyTo` and `snapshotsCopyFrom` paths to the TimescaleDB container, data node will panic when trying to start from network history because TimescaleDB is not able to access the network history files that data node retrieves and tries to import. 
:::
```

- `database_password` is the password you want to use to connect to the database.
- `database_name` is the name of the database you want to use for storing the data.
- `localdb_port` is the port you want to use to connect to the database on your local machine. (5432 is the default port for Postgresql database server and may not be available if you already have a postgresql database server running on your machine and want to use Docker for testing).

> - `/host_path/to/snapshotsCopyTo` is the path on your host machine where you want to store the snapshots that are generated by data node. Snapshots allow you to restore the database to a previous state.

Reviewer 2: Sort of gives the impression that this is where we store the network history, when really they are temp working used when creating/loading history segments, however, I think the change Phil is making will remove the need for these paths, so perhaps not worth updating this for now.

The following will wipe the database on startup, so use with caution.

Once the database has been wiped, data node will reconstruct the database tables and will allow you to repopulate the data from the network chain, however this can take a long time depending on the size of the chain.

> Currently the default value for `WipeOnStartup` is `true`. Please ensure this is updated if you do not want the data to be wiped when Data Node starts.

Reviewer: ```suggestion
Currently the default value for `WipeOnStartup` is `true`. Update this if you do not want all data to be wiped when data node starts.
```

This will cause data node to download a specially prepared Postgresql package which is extracted to your local machine if it doesn't exist. A separate Postgresql process will be spawned by data node using the credentials you specified in the database configuration section. Once data node is stopped, the child Postgresql process will be stopped automatically.

> You can launch PostgreSQL in its own separate process using the data node embedded postgresql binaries by running the following command:

Reviewer: ```suggestion
You can launch PostgreSQL in its own separate process using the data node embedded PostgresQL binaries by running the following command:
```

In either case, the files for the database will be stored in the data node `state` folder located at `YOUR_DATA_NODE_HOME_PATH/state.data-node/storage/sqlstore`.

> ### Buffered event source

Reviewer 2: Does this need to be in the doc I wonder?  there's a lot of info in the doc, the defaults for this are generally fine, anything we can do to keep the doc simple as possible is probably good as long as we think its not important for the operator to know about it, this is almost implementation detail, so I think falls into that category, thoughts?

Author: I also think it is helpful for the end user in case they need to troubleshoot.

Reviewer 2: it creates the error in the state directory, which must have write perms for other things, and will create the dir at startup if not exists, so should be fine

## Fetch network history
After starting a data node, you can load in a segment of network history, if you want your node to have more data than provided by the current height. This is particularly useful if you're running an archive node.

> Before you can fetch network history, you will need to update your data node configuration file adding a list of bootstrap peers as by default this is left empty and depends on which network you are connecting to:

Reviewer: ```suggestion
Before you can fetch network history, you will need to update your data node configuration file to add a list of bootstrap peers, as by default this is left empty and depends on which network you are connecting to:
```

## Data Node recovery from Network History

It is possible to start a data node and initialize it using network history automatically. By default, this process is disabled, and if enabled will only fetch - if you don't have the snapshot locally - the last segment before loading it into the data node. This is to allow users to quickly initialize a data node without fetching a lot of data and start participating while the data node is fetching more data in the background.

To enable this feature set the `AutoInitialiseFromNetworkHistory` setting in the data node configuration file to `true`. It is safe to leave this setting to true. If `WipeOnStartup` is `true`, data node will always restart with the last network history it is able to retrieve. If `WipeOnStartup` is `false`, data node will not reload data if the block height available in the database is equal to or greater than the block height from the network history.

If you want to fetch more than the last segment, you may also set the `MinimumBlockCount` configuration setting.

> If you want to initialise the data node automatically up to a specific segment and X blocks before it, you can set both the `ToSegment` and `MinimumBlockCount` configuration settings.

Reviewer 3: Can We add information here on what timeouts We must update?

Author: Resolved https://github.com/vegaprotocol/documentation/pull/556/commits/7c42130487b2326cabd0e9470ba7f297376f173a

## Fetch network history
After starting a data node, you can load in a segment of network history, if you want your node to have more data than provided by the current height. This is particularly useful if you're running an archive node.

Before you can fetch network history, you will need to update your data node configuration file to add a list of bootstrap peers, as by default this is left empty and depends on which network you are connecting to:

```toml
[NetworkHistory]
...
  [NetworkHistory.Store]
  ...
  BootstrapPeers = ["/path/to/bootstrap-peer/1","/path/to/bootstrap-peer/2"]
```

> To get a list of bootstrap peers available for your network, you can make a HTTP request to the API service for your chosen network, for example: `https://api.stagnet1.vega.xyz/api/v2/networkhistory` will retrieve the network peers for `stagnet1`.

Reviewer: Could we use testnet/fairground here as it's a network someone external could try this with?

```json
{
  "ipfsAddress": "/ip4/<data-node-ip-4-address>/tcp/4001/p2p/<some-uid>",
  "swarmKey": "/key/swarm/psk/1.0.0/\n/base16/\n<some key>",
  "swarmKeySeed": "<some-key-seed>",
  "connectedPeers": [
    "/ip4/<peer-ip-4-address>/tcp/4001/p2p/<peer-id>",
    ...
  ]
}
```

To see how much network history your data node has, run the following command:

```shell
vega datanode network-history show --home="YOUR_DATA_NODE_HOME_PATH"
```

> If you are building a new data node, you may have no history and you will see this message:

Reviewer: I think this could use a sub-section title, for easier navigation

...
```

:::note
It is possible that some data peers are behind and may not have the latest data. However, the `history_segment_id` should be the same across all nodes that are at the same height.
:::

To fetch a network history segment, run the command below. Use the ID of the segment you want (for example, the oldest) followed by the number of blocks prior to the segment's height that you want fetch. `2000` is used in the following example. This will result in all blocks from height 3000 to 5000 being retrieved.

```shell
vega datanode network-history fetch <segment-id-of-segment-at-height-X> 2000 --home="YOUR_DATA_NODE_HOME_PATH"
```

Once the network history segments have been downloaded, running: 

```shell
vega datanode network-history show --home="YOUR_DATA_NODE_HOME_PATH"
``` 

should display the network history you have:

```text
Available contiguous history spans:

Contiguous history from block height XXXXX to XXXXX, from segment id: <first-segment-id> to <last-segment-id>

Datanode contains no data
```

> ### Loading network history into Data Node

Reviewer: ```suggestion
### Loading network history into data node
```

Now that you have downloaded historical data you can load the history into the data node using:

```shell
vega datanode network-history load --home="YOUR_DATA_NODE_HOME_PATH"
```

:::note
Data node must not be running when you use this command. If you attempt to use `network-history load` while data node is running, you will get an error:

```text
datanode must be shutdown before data can be loaded
```
:::

You will be notified that the load command will force all existing connections to the data node database will be closed automatically and prompted if you want to continue. Enter `y` and data node will check how much history you have, tell you what it will load, and prompts you to confirm if you want to continue. Enter `y` again and the data node will load the history you have fetched.

This process may take a very long time depending on how much history you have retrieved and are loading into the database. Some processes in the restoration can take a long time to complete and it may look like the process is not doing anything. Do not try to terminate the process early or you may end up with a corrupted data node database.

> 
> ## Data Node recovery from Network History

Reviewer: ```suggestion
## Data node recovery from Network History
```

> It is possible to start a data node and initialize it using network history automatically. By default, this process is disabled, and if enabled will only fetch - if you don't have the snapshot locally - the last segment before loading it into the data node. This is to allow users to quickly initialize a data node without fetching a lot of data and start participating while the data node is fetching more data in the background.

Reviewer: ```suggestion
It is possible to start a data node and initialise it using network history automatically. By default, this process is disabled, and if enabled will only fetch - if you don't have the snapshot locally - the last segment before loading it into the data node. This is to allow users to quickly initialise a data node without fetching a lot of data and start participating while the data node is fetching more data in the background.
```

> To enable this feature set the `AutoInitialiseFromNetworkHistory` setting in the data node configuration file to `true`. It is safe to leave this setting to true. If `WipeOnStartup` is `true`, data node will always restart with the last network history it is able to retrieve. If `WipeOnStartup` is `false`, data node will not reload data if the block height available in the database is equal to or greater than the block height from the network history.

Reviewer: ```suggestion
To enable this feature, set the `AutoInitialiseFromNetworkHistory` setting in the data node configuration file to `true`. It is safe to leave this setting to true. 

If `WipeOnStartup` is `true`, data node will always restart with the last network history it is able to retrieve. If `WipeOnStartup` is `false`, data node will not reload data if the block height available in the database is equal to or greater than the block height from the network history.
```

    "/ip4/<peer-ip-4-address>/tcp/4001/p2p/<peer-id>",
    ...
  ]
}
```

To see how much network history your data node has, run the following command:

```shell
vega datanode network-history show --home="YOUR_DATA_NODE_HOME_PATH"
```

### Fetching network history for a new data node

If you are building a new data node, you may have no history and you will see this message:

```shell
No network history is available. Use the fetch command to fetch network history
```

In this case, it is possible to get the latest history segment ids from your network history bootstrap peers:

```shell
vega datanode network-history latest-history-segment-from-peers --home="YOUR_DATA_NODE_HOME_PATH"
```

This will return a list of peers you can fetch network history from:

```text
Most Recent History Segments:

Peer:<peer-ip-and-api-port>, Swarm Key:<swarm-key>, Segment{from_height:75001  to_height:75300  history_segment_id:"some-segment-id"  previous_history_segment_id:"some-other-segment-id"}
...
```

:::note
It is possible that some data peers are behind and may not have the latest data. However, the `history_segment_id` should be the same across all nodes that are at the same height.
:::

> To fetch a network history segment, run the command below. Use the ID of the segment you want (for example, the oldest) followed by the number of blocks prior to the segment's height that you want fetch. `2000` is used in the following example. This will result in all blocks from height N-2000 to N being retrieved.

Reviewer 2: ```suggestion
To fetch a network history segment, run the command below. Use the ID of the segment you want (for example, the latest) followed by the number of blocks prior to the segment's height that you want fetch. `2000` is used in the following example. This will result in all blocks from height N-2000 to N being retrieved.
```

```shell
vega datanode network-history show --home="YOUR_DATA_NODE_HOME_PATH"
``` 

should display the network history you have:

```text
Available contiguous history spans:

Contiguous history from block height XXXXX to XXXXX, from segment id: <first-segment-id> to <last-segment-id>

Datanode contains no data
```

### Loading network history into data node
Now that you have downloaded historical data you can load the history into the data node using:

```shell
vega datanode network-history load --home="YOUR_DATA_NODE_HOME_PATH"
```

:::note
Data node must not be running when you use this command. If you attempt to use `network-history load` while data node is running, you will get an error:

```text
datanode must be shutdown before data can be loaded
```
:::

You will be notified that the load command will force all existing connections to the data node database will be closed automatically and prompted if you want to continue. Enter `y` and data node will check how much history you have, tell you what it will load, and prompts you to confirm if you want to continue. Enter `y` again and the data node will load the history you have fetched.

> This process may take a very long time depending on how much history you have retrieved and are loading into the database. Some processes in the restoration can take a long time to complete and it may look like the process is not doing anything. Do not try to terminate the process early or you may end up with a corrupted data node database.

Reviewer 2: should have fixed the corrupted data node problem, in current PR tbc

Author: I'll leave it in for now, better to err on the side of caution.

To enable this feature, set the `AutoInitialiseFromNetworkHistory` setting in the data node configuration file to `true`. It is safe to leave this setting to true. 

If `WipeOnStartup` is `true`, data node will always restart with the last network history it is able to retrieve. If `WipeOnStartup` is `false`, data node will not reload data if the block height available in the database is equal to or greater than the block height from the network history.

If you want to fetch more than the last segment, you may also set the `MinimumBlockCount` configuration setting.

If you want to initialise the data node automatically up to a specific segment and X blocks before it, you can set both the `ToSegment` and `MinimumBlockCount` configuration settings.

```toml
AutoInitialiseFromNetworkHistory = true

[NetworkHistory.Initialise]

>   ToSegment = "<segment-id-of-last-segment-you-require>" 

Reviewer 3: How can we get list of segments (or at least the last one) once we do not have data node running?

Author: If you leave it empty, it will by default fetch the latest segment. You would only populate this field if you need to initialise a data-node upto a specific segment, for example if you want to spin up a data node to get reporting data for a specific period.

I will update the docs to mention this specifically if it's not there.

## Further reading
For more information about data node and developing on data node please see the data node [README ↗](https://github.com/vegaprotocol/vega/blob/master/datanode/README.md)

## Data node troubleshooting

### Block height on begin block is too high

If you start data node and you receive an error that looks like:

```text
block height on begin block, XXXXXX, is too high, the height of the last processed block is XXXXXX
```

> This indicates that the core node is out of sync with the data node it is connected to.

Reviewer 2: ```suggestion
This indicates that the core node block height is ahead of the data node it is connected to.
```
